0000-modules/overview.md

```diff
@@ -0,0 +1,190 @@
+paths. These paths look exactly like they look today:
+
+```rust
+use std::collections::HashMap;
+use serde::Serialize;
+::std::iter::Iterator
+::serde::de::DeserializeOwned
+```
+
+The major difference from today is that `extern crate` declarations are no
+longer necessary. This change however is covered in [RFC #2088][extern-crate],
+not directly by this RFC.
+
+### Adding a new file to your crate
+
+When a user wants to break their project into multiple files, they can do so by
+creating a new `.rs` file in the directory that contains their crate root (or a
+subdirectory thereof). For most users, this would be the `src` directory.
+
+This file will automatically be picked up to create a new **module** of the
+user's crate. The precise mechanism is discussed in [another
+section][loading-files] of the RFC. In brief, cargo will find the file and tell
+rustc to treat it as a new module using an interface that is also available to
+non-cargo users. Cargo users who do not want this behavior to be automatic can
+instead specify their modules using a field in their `Cargo.toml`.
+
+The major difference from today is that a `mod` statement is no longer
+necessary to get rustc to look up and load your files. The `mod` keyword still
+exists to support inline modules (`mod foo { ... }`).
+
+### Making items public to the rest of your crate
+
+By default, all modules are public to the rest of the crate. However, all
+items inside them - like types, functions, traits, etc - are private to that
+module. To make them public to the rest of the crate, the can use the `pub`
+keyword:
+
+```rust
+// Private to this module
+struct Foo;
+
+// Public to the crate
+pub struct Bar;
+```
+
+Users can make it public to only part of your crate using the `pub(restricted)`
+syntax that already exists.
+
+There are two significant differences from today:
+
+* All modules which are automatically loaded are public to the rest of the
+crate, rather than taking a visibility modifier.
+* The `pub` keyword means *public to this crate* - the equivalent of today's
+`pub(crate)`. A new keyword is introduced to mean a part of the public API;
+this is discussed in a subsequent section.
+
+### Importing items from other parts of your crate
+
+Once a user has a `pub` item in another module, they can import it using the
+`use` statement, just like items from external dependencies. Paths to items in
+modules of this crate work like this:
+
+* All these paths begin with the `crate` keyword, which means "this crate."
+* All modules in the `src` directory are mounted in the root module, all
+modules in other directories are mounted inside the module for that directory.
+* Items are mounted in the module that defines them.
+
+So if a user has a structure like this:
+
+```
+// src/foo.rs
+pub struct Bar;
+```
+
+They can access it with `use` and absolute paths like this:
+
+```rust
+use crate::foo::Bar;
+::crate::foo::Bar;
```

Comment:
So in each module, the root `::` contains external crates, this crate, this module, and this module's patent? One nice thing about this is that it means you could say "`use` paths are always relative to `::`", rather than needing the exception "unless they begin with `crate`, `self`, or `super`".